presto-main/src/main/java/com/facebook/presto/operator/TableFinishOperator.java

```diff
@@ -136,16 +177,62 @@ public void addInput(Page page)
                 fragmentBuilder.add(VARBINARY.getSlice(fragmentBlock, position));
             }
         }
+
+        Optional<Page> statisticsPage = extractStatisticsRows(page);
+        statisticsPage.ifPresent(statisticsAggregation::addInput);
+    }
+
+    private static Optional<Page> extractStatisticsRows(Page page)
+    {
+        int statisticsPositionCount = 0;
+        for (int position = 0; position < page.getPositionCount(); position++) {
+            if (isStatisticsPosition(page, position)) {
+                statisticsPositionCount++;
+            }
+        }
+
+        if (statisticsPositionCount == 0) {
+            return Optional.empty();
+        }
+
+        if (statisticsPositionCount == page.getPositionCount()) {
+            return Optional.of(page);
+        }
+
+        int selectedPositionsIndex = 0;
+        int[] selectedPositions = new int[statisticsPositionCount];
+        for (int position = 0; position < page.getPositionCount(); position++) {
+            if (isStatisticsPosition(page, position)) {
+                selectedPositions[selectedPositionsIndex] = position;
+                selectedPositionsIndex++;
+            }
+        }
+
+        Block[] blocks = new Block[page.getChannelCount()];
+        for (int channel = 0; channel < page.getChannelCount(); channel++) {
+            blocks[channel] = page.getBlock(channel).copyPositions(selectedPositions, 0, statisticsPositionCount);
```

Comment:
Copy just seems to be safer. Because it is copy. There will be not that much data to copy. And in most of the cases it will return the page unaltered. But it can be `get` as well. I don't have a strong opinion here.

Reply:
Well -- we sometimes return the whole page (few lines earlier), so there cannot be assumption that you do a copy.

(no strong opinion either, just opportunity to simplify the code)